[MRESOLVER-499] IPC Named locks

cstamas opened the pull request:
Implementation of IPC Named Locks. Code originates from mvnd, this was the original locking it used, adapted to NamedLocks, that now became possible.

---

https://issues.apache.org/jira/browse/MRESOLVER-499

cstamas commented on maven-resolver-named-locks-ipc/pom.xml:
```diff
@@ -0,0 +1,118 @@
+  <description>A synchronization utility implementation using IPC.</description>
+
+  <properties>
+    <javaVersion>9</javaVersion>
```
This is due com.sun stuff, any ideas?

michael-o replied:
UDS in 16 was incomplete, that was fixed in 17. A great alternative is rather https://github.com/kohlschutter/junixsocket. Would consider that one to reach a broader audience.

cstamas replied:
You are right: we should go for 17 then, basically reduce our "resolution" to LTS versions only. The Artifact Generator PR uses UDS and is currently 16... This one has non trivial dependency trail (and native stuff).

cstamas replied:
Also, just figured: `sun.misc.Signal ` is unrelated to UDS...

olamy replied:
Another maintained option, Jetty have been using for long time
```
      <groupId>com.github.jnr</groupId>
      <artifactId>jnr-unixsocket</artifactId>
```

michael-o commented on maven-resolver-named-locks-ipc/src/main/java/org/eclipse/aether/named/ipc/IpcMessages.java:
```diff
@@ -0,0 +1,36 @@
+    public static final String RESPONSE_ACQUIRE = "response-acquire";
+    public static final String RESPONSE_CLOSE = "response-close";
+    public static final String RESPONSE_STOP = "response-stop";
+}
```
I wonder wether an enum would be better?!

gnodet replied:
It would require rewriting the socket protocol, which is currently written to read/write a `List<String>`, so definitely not an enum:
https://github.com/apache/maven-resolver/blob/814b6732a49447281ef085de3ef8542c45508371/maven-resolver-named-locks-ipc/src/main/java/org/eclipse/aether/named/ipc/IpcClient.java#L266-L308

michael-o commented on maven-resolver-named-locks-ipc/src/main/java/org/eclipse/aether/named/ipc/SocketFamily.java:
```diff
@@ -0,0 +1,111 @@
+ * @since 2.0.0
+ */
+public enum SocketFamily {
+    inet;
```
Should be uppercase. Akin to `AF_INET`.

michael-o commented on maven-resolver-named-locks-ipc/src/main/java/org/eclipse/aether/named/ipc/SocketFamily.java:
```diff
@@ -0,0 +1,111 @@
+    }
+
+    public static SocketAddress fromString(String str) {
+        if (str.startsWith("inet:")) {
```
Why this complexity?

gnodet replied:
Because UDS was supported in the original version:
https://github.com/apache/maven-mvnd/blob/0.7.1/common/src/main/java/org/mvndaemon/mvnd/common/SocketFamily.java

gnodet replied:
Maybe adding the UDS support and the unit test would help: https://github.com/apache/maven-mvnd/blob/0.7.1/common/src/test/java/org/mvndaemon/mvnd/common/SocketFamilyTest.java

michael-o commented on maven-resolver-named-locks-ipc/src/main/java/org/eclipse/aether/named/ipc/IpcClient.java:
```diff
@@ -0,0 +1,417 @@
+
+        Path lockPath = this.lockPath.toAbsolutePath().normalize();
+        Path lockFile =
+                lockPath.resolve(".maven-resolver-ipc-lock-" + family.name().toLowerCase());
```
`Locale.ROOT`

michael-o commented on maven-resolver-named-locks-ipc/src/main/java/org/eclipse/aether/named/ipc/IpcClient.java:
```diff
@@ -0,0 +1,417 @@
+
+    SocketChannel createClient() throws IOException {
+        String familyProp = System.getProperty(IpcServer.SYSTEM_PROP_FAMILY, IpcServer.DEFAULT_FAMILY);
+        SocketFamily family = familyProp != null ? SocketFamily.valueOf(familyProp) : SocketFamily.inet;
```
How can this `null` if you have a default value?

michael-o commented on maven-resolver-named-locks-ipc/src/main/java/org/eclipse/aether/named/ipc/IpcServer.java:
```diff
@@ -0,0 +1,482 @@
+        try {
+            sun.misc.Signal.handle(new sun.misc.Signal("INT"), sun.misc.SignalHandler.SIG_IGN);
+            if (IpcClient.IS_WINDOWS) {
+                sun.misc.Signal.handle(new sun.misc.Signal("TSTP"), sun.misc.SignalHandler.SIG_IGN);
```
I don't understand this because Windows does not use signals at all.

michael-o commented on maven-resolver-named-locks-ipc/src/main/java/org/eclipse/aether/named/ipc/IpcServer.java:
```diff
@@ -0,0 +1,482 @@
+
+    private static void debug(String msg, Object... args) {
+        if (DEBUG) {
+            System.out.printf("[ipc] [debug] " + msg + "\n", args);
```
`%n`

michael-o commented on maven-resolver-named-locks-ipc/src/main/java/org/eclipse/aether/named/ipc/IpcServer.java:
```diff
@@ -0,0 +1,482 @@
+    }
+
+    private static void info(String msg, Object... args) {
+        System.out.printf("[ipc] [info] " + msg + "\n", args);
```
`%n`

cstamas commented:
IPC code was lifted from @gnodet PR :smile: so will let him answer.

olamy commented:
just saying in case you don't want to reinvent the wheel. You can server/client unixsocket already done in Jetty https://github.com/jetty/jetty.project/tree/jetty-10.0.x/jetty-unixsocket

cstamas commented:
There is a mixup here: UDS is NOT used here at all, it is used by other PR, the Signer to communicate with gpg-agent. This PR does not use UDS, it uses plain inet sockets, which IMO is okay, as that could mean that Server could be started even on remote host (did not test, just guessing).

slawekjaranowski commented on maven-resolver-named-locks-ipc/src/main/java/org/eclipse/aether/named/ipc/IpcClient.java:
```diff
@@ -0,0 +1,417 @@
+                ServerSocketChannel ss = family.openServerSocket();
+                String tmpaddr = SocketFamily.toString(ss.getLocalAddress());
+                String rand = Long.toHexString(new Random().nextLong());
+                String syncCmd = IS_WINDOWS ? "mvnd-sync.exe" : "mvnd-sync";
```
do we need external executables?

cstamas replied:
Original mvnd code did use graalvm to make native image out of server, that stuff I did not bring over, but we may...

gnodet commented:
> just saying in case you don't want to reinvent the wheel. You can server/client unixsocket already done in Jetty https://github.com/jetty/jetty.project/tree/jetty-10.0.x/jetty-unixsocket

The whole idea was to have a minimal set of dependencies so that it could easily be turned into a native executable...

cstamas commented:
PR updated to latest Resolver master. If anyone wants to pick up, go for it. Otherwise I may pick it up later.

cstamas commented:
This PR now updates IPC locks, added unix socket family (as module was already Java 17!), added UT, default made `unix` and the ITs run with it as well.

jira-importer commented:
Resolve #1171
